pkg/che/tokencache.go

```diff
@@ -0,0 +1,153 @@
+package che
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	crtCfg "github.com/codeready-toolchain/member-operator/pkg/configuration"
+	"github.com/codeready-toolchain/member-operator/pkg/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/pkg/errors"
+)
+
+const tokenPath = "auth/realms/che/protocol/openid-connect/token"
+
+// tokenCache manages retrieving, caching and renewing the authentication token required for invoking Che APIs
+type tokenCache struct {
+	sync.RWMutex
+	httpClient *http.Client
+	token      *TokenSet
+}
+
+func newTokenCache() *tokenCache {
+	return &tokenCache{
+		httpClient: newHTTPClient(),
+	}
+}
+
+// getToken returns the token needed to use Che user APIs, or an error if there was a problem getting the token
+func (tc *tokenCache) getToken(cl client.Client, cfg *crtCfg.Config) (TokenSet, error) {
+	defer tc.RUnlock()
```

Comment:
This function not just reads the token. It also writes it in line 62: `obtainNewToken()`. So you can't just use RLock here.
You can use something like this: https://github.com/codeready-toolchain/devcluster/blob/master/pkg/ibmcloud/cluster.go#L64

Reply:
oops, I actually meant to use `Unlock` and `Lock` not `RUnlock` and `RLock`. Let me try something like https://github.com/codeready-toolchain/devcluster/blob/master/pkg/ibmcloud/cluster.go#L64. I did briefly try that but ran into some issue, but let me try again.

Reply:
Fixed in 24cf6c55f1de637ac561cbd681c6896adb87fa6a